### ISSUE TITLE
Make localAddress/remoteAddress mockable in EmbeddedChannel

### ISSUE THREAD

**Issue by mkieselmann** [apple/swift-nio#829]
### Problem
Testing channel handlers that rely on `Channel.localAddress` or `Channel.remoteAddress` is currently quite inconvenient as one cannot set test values for `localeAddress` or `remoteAddress` on `EmbeddedChannel`.

### Expected behavior

1) The two properties could be variables instead of constants.
```
public class EmbeddedChannel: Channel {
    // ...
    public var localAddress: SocketAddress? = nil
    public var remoteAddress: SocketAddress? = nil
    // ...
}
```

2) `EmbeddedChannel` could be `open` for subclassing. 
```
open class EmbeddedChannel: Channel {
    // ...
    open let localAddress: SocketAddress? = nil
    open let remoteAddress: SocketAddress? = nil
    // ...
}
```

### Actual behavior
 It seems that the only solution is to replicate most of the code of `EmbeddedChannel` and implement `ChannelHandler` protocol.

**Comment by weissi** [comment]
@mkieselmann thanks for your report! One other option we would have is just to add a setter on `EmbeddedChannel`, right? That way you could just set them which would also be enough, right?

We could also just store the values from `connect` and `bind` but that's a bit odd too...

@Lukasa  what do you think?

**Comment by mkieselmann** [comment]
@weissi thanks for the quick response. Sure, having a setter would be fine. Not sure how it would work with `connect` or `bind` since we would need some kind of bootstrapping in the test setup. Currently I just have an `EmbeddedChannel` with one or more `ChannelHandlers` added to the pipeline.

**Comment by Lukasa** [comment]
@weissi If we were going to have a setter, why not just make them `var` properties?

**Comment by weissi** [comment]
> @weissi If we were going to have a setter, why not just make them `var` properties?

Yes, that works and it a really easy one to PR :)

**Comment by weissi** [comment]
> Not sure how it would work with `connect` or `bind` since we would need some kind of bootstrapping in the test setup.

You could just

XCTAssertNoThrow(channel.connect(...).wait())
XCTAssertNoThrow(channel.bind(...).wait())

so you don't need to use a bootstrap if you're not actually doing a real connection. But I think the setters (through `var localAddress` / var `remoteAddress`) seem better anyway.

**Comment by weissi** [comment]
we can obviously also do both, set the addresses from `connect`/`bind` as well as allowing the properties to be set directly...

**Comment by mkieselmann** [comment]
> You could just
>
> ```swift
> XCTAssertNoThrow(channel.connect(...).wait())
> XCTAssertNoThrow(channel.bind(...).wait())
> ```
> 
> so you don't need to use a bootstrap if you're not actually doing a real connection. But I think the setters (through `var localAddress` / var `remoteAddress`) seem better anyway.
thanks makes sense :-)

> we can obviously also do both, set the addresses from `connect`/`bind` as well as allowing the properties to be set directly...
I created a PR for that: #833

**Comment by weissi** [comment]
fixed with #833, thanks @mkieselmann
